### ISSUE TITLE
[fbpick-coarse] Add minimal E2E smoke test for global-anchor coarse inference

### ISSUE THREAD

**Issue by ozw4** [ozw4/seisai#45]
Background

fbpick-coarse は global-anchor resize mode へ移行した。

現在の coarse inference flow は以下。

raw SEG-Y
→ full gather sampling
→ gap-aware center anchor selection
→ fixed input: (3, 256, 2048)
→ model forward
→ anchor prediction
→ original trace/sample coordinate restoration
→ .coarse.npz save

この pipeline は複数の処理をまたぐため、unit test だけでは以下の接続が壊れていないことを確認しづらい。

config loading
dataset construction
model loading
inference loop
coordinate restoration
.coarse.npz writing

この issue では、速度比較や memory benchmark は行わず、global-anchor coarse inference が最小構成で E2E 実行できることを確認する smoke test を追加する。

Goal

小さな synthetic / fixture データと dummy model を使って、fbpick-coarse の global-anchor inference が end-to-end で完走し、.coarse.npz を出力できることを確認する。

この issue の目的は performance 測定ではない。

Do:
  - minimal E2E smoke test
  - output file existence check
  - output schema / shape check

Do not:
  - speed benchmark
  - memory benchmark
  - AMP comparison
  - num_workers comparison
  - throughput comparison
Scope
1. Add minimal E2E inference smoke test

対象候補:

packages/seisai-engine/tests/test_fbpick_coarse.py

または E2E 用に分けるなら:

packages/seisai-engine/tests/test_fbpick_coarse_e2e.py

test は以下を実行する。

1. small synthetic SEG-Y fixture を用意する
2. global-anchor coarse infer config を作る
3. dummy checkpoint / dummy model を用意する
4. run_coarse_infer または CLI 相当を実行する
5. .coarse.npz が作成されることを確認する
6. .coarse.npz の required fields / shapes を確認する
2. Use the smallest practical fixture

test は速く・安定している必要がある。

fixture 方針:

- n_traces は 256 未満でもよい
- n_samples は 2048 未満または少し大きい程度でよい
- pad row path も自然に通るとよい
- real SEG-Y fixture が既存 test helper で作れるならそれを使う

例:

n_traces = 32
n_samples = 256
dt_sec = 0.001
primary key = ffid
secondary key = chno

ただし、global-anchor input は内部で必ず以下になる。

input shape: (3, 256, 2048)
3. Use a deterministic dummy model

本 test の目的は model accuracy ではないため、学習済み model は不要。

dummy model は以下で十分。

input:  (B, 3, 256, 2048)
output: (B, 1, 256, 2048)

例えば、time channel を logits に使う model にすると deterministic になる。

class DummyCoarseModel(torch.nn.Module):
    def forward(self, x):
        # x[:, 2:3] is time_ch
        return x[:, 2:3]

ただし checkpoint metadata validation を通す必要がある場合は、global-anchor coarse 用 metadata を含む checkpoint を test 内で作る。

必要 metadata:

coarse_input_mode = "global_anchor_resize"
coarse_trace_len = 256
coarse_time_len = 2048
coarse_in_chans = 3
coarse_input_channels = ["waveform", "offset_ch", "time_ch"]
4. Run global-anchor coarse inference

test では、可能なら public entrypoint に近いところを通す。

優先順位:

1. run_coarse_infer(...)
2. cli/run_fbpick_coarse_infer.py の main 相当
3. 既存 test helper 経由

E2E smoke test なので、内部 helper だけを直接呼ぶのではなく、なるべく実際の inference path に近い経路を通す。

ただし、CLI subprocess まで使うと重くなる場合は、Python 関数呼び出しでよい。

5. Validate .coarse.npz output contract

出力 .coarse.npz を読み、最低限の field と shape を確認する。

required fields 例:

coarse_pick_i
coarse_pick_t_sec
coarse_pmax
dt_sec
n_samples_orig
n_traces
trace_indices
offsets_m

期待:

coarse_pick_i.shape == (n_traces,)
coarse_pick_t_sec.shape == (n_traces,)
coarse_pmax.shape == (n_traces,)

さらに以下も確認する。

coarse_pick_i is integer
0 <= coarse_pick_i < n_samples_orig
coarse_pick_t_sec == coarse_pick_i * dt_sec
coarse_pmax is finite

np.allclose でよい。

6. Ensure legacy tiled path is not used

global-anchor smoke test では legacy tiled inference が呼ばれないことも確認したい。

可能なら monkeypatch で以下が呼ばれたら fail するようにする。

iter_infer_loader_tiled_w
TiledWConfig
InferenceGatherWindowsDataset

ただし、既に import が消えているならこの test は不要。
代わりに config parser が legacy key を reject する既存 test に任せてよい。

Out of scope

この issue では以下は行わない。

- speed benchmark
- memory benchmark
- GPU memory measurement
- CPU memory measurement
- throughput report
- AMP on/off comparison
- num_workers comparison
- batch_size comparison
- profiling
- visualization
- model accuracy evaluation
- coverage@fine-window evaluation
Acceptance criteria
global-anchor coarse inference の minimal E2E smoke test が追加されている。
test は small synthetic / fixture SEG-Y で実行できる。
test は dummy model / dummy checkpoint を使い、学習済み checkpoint を必要としない。
inference path が完走する。
.coarse.npz が出力される。
.coarse.npz に required fields が存在する。
coarse_pick_i.shape == (n_traces,)。
coarse_pick_t_sec.shape == (n_traces,)。
coarse_pmax.shape == (n_traces,)。
coarse_pick_t_sec == coarse_pick_i * dt_sec の整合性が確認される。
test は performance threshold を持たない。
test は速度や memory の数値を assertion しない。
既存 tests が通る。
Suggested test name
def test_fbpick_coarse_global_anchor_infer_e2e_smoke(...):
    ...

または、

def test_global_anchor_coarse_inference_writes_npz_smoke(...):
    ...
Suggested minimal assertions
out_files = list(out_dir.glob("*.coarse.npz"))
assert len(out_files) >= 1

with np.load(out_files[0]) as z:
    coarse_pick_i = z["coarse_pick_i"]
    coarse_pick_t_sec = z["coarse_pick_t_sec"]
    coarse_pmax = z["coarse_pmax"]
    dt_sec = float(z["dt_sec"])
    n_traces = int(z["n_traces"])
    n_samples_orig = int(z["n_samples_orig"])

assert coarse_pick_i.shape == (n_traces,)
assert coarse_pick_t_sec.shape == (n_traces,)
assert coarse_pmax.shape == (n_traces,)
assert np.issubdtype(coarse_pick_i.dtype, np.integer)
assert np.all(coarse_pick_i >= 0)
assert np.all(coarse_pick_i < n_samples_orig)
assert np.all(np.isfinite(coarse_pick_t_sec))
assert np.all(np.isfinite(coarse_pmax))
np.testing.assert_allclose(
    coarse_pick_t_sec,
    coarse_pick_i.astype(np.float32) * np.float32(dt_sec),
    rtol=0,
    atol=1.0e-6,
)
